### ISSUE TITLE
Model Training for Python code vectors

### ISSUE THREAD

**Issue by SimoneBrigante** [tech-srl/code2vec#71]
Hi, 
I would like to thank you for the really good job you have done.
I am interested in using code2vec to obtain an array representation of some python methods and then use them, eventually together with other features, in a binary classification model.
Since I just want the arrays I don’t know how to approach the training phase.
Should I train your model using as label the method name, as you do, or should I already change it to the boolean values?
Because, since I don't have a huge dataset, I was wondering if I could train the model on a larger dataset found online and then use the trained model to get the arrays of the methods I am interested in.

Thanks in advance
Simone

**Comment by urialon** [comment]
Hi Simone,
Thank you for your interest in code2vec and for your kind words.

It's hard to tell. One option would be to train the model using the boolean labels, this will optimize your objective directly. With some additional work, you can also use your additional features and feed them into training.

If your dataset is too small, it might be useful to train a model to predict method names on a larger dataset, and then stop, and fine-tune the model using the binary classification labels.

The option you are suggesting, of training for method names, export the vectors, and feed them into another binary classifier - also sounds plausible.

It's difficult to tell in advance... I suggest that you try all options, it will be interesting to see a comparison of these approaches.

Best,
Uri

P.S - our newer model [code2seq](https://code2seq.org/) performs better than code2vec, even when the output is a label and not a sequence.

**Comment by SimoneBrigante** [comment]
> It's hard to tell. One option would be to train the model using the boolean labels, this will optimize your objective directly. With some additional work, you can also use your additional features and feed them into training.

The idea of including the other features directly into the training phase seems interesting, do you have any suggestion on how it could be done or where I should look in the code?

> If your dataset is too small, it might be useful to train a model to predict method names on a larger dataset, and then stop, and fine-tune the model using the binary classification labels.

My dataset will contain around 15k methods, they are extracted from students programming exercises, so I will have many implementations of the same methods.
What do you mean by "fine-tune" the model? How could I do it?
The main problem here would probably be the training of the large dataset, do you know, by chance, if someone has already trained a model for python on a large dataset?

> P.S - our newer model [code2seq](https://code2seq.org/) performs better than code2vec, even when the output is a label and not a sequence.

I was already considering to try it too, since I believe that the preprocessing part is the same as the code2vec

Sorry for the many questions and thank you for your help!
Simone

**Comment by urialon** [comment]
Including the other features requires to add them to the dataset file.

Please create a temporary dataset file, and examine its structure: every line is an example; the first "word" is the target label that needs to be predicted, and the rest of the 200 "contexts" are the paths that are the input of the network.

The next challenge would be to propagate them from the text file, through the "reader", up to the model itself.
I think the easiest way will be to add these features, space-separated, between the label and the path.
Then, extend the csv_record_default (currently it has 201 items, you need it to have 201 + number_of_features here:
https://github.com/tech-srl/code2vec/blob/master/path_context_reader.py#L82

Then, hopefully you'll get the features as the elements 1:num_features in `row_parts` here:
https://github.com/tech-srl/code2vec/blob/master/path_context_reader.py#L185

And make sure that `context_str` are taken from the correct indices here (add num_features to the beginning and end of the current indexing):
https://github.com/tech-srl/code2vec/blob/master/path_context_reader.py#L189

Then, follow the code and see how the rest of the inputs (such as the path contexts) flow into the model.

Regarding training a Python model - I don't know. You might want to check previous closed issues, because some people have asked about python in the past.

Best,
Uri

**Comment by SimoneBrigante** [comment]
Hi Uri,
Thanks for your detailed reply, I will try to add my features too.

Until now I have done some testing using just path contexts, 
I have tried to train a code2vec model both for method name prediction and binary classification.
The first one reaches really high scores, but probably because the methods are extracted from simple programming exercises, so most of them are similar.
In the second case the scores remain constant during epochs.
Both models stop training after 18 epochs, is it normal?
As expected when I extract the vector representations and use them in my model I obtain better performances from the ones of the binary classification model.

I attach the outputs of both training sessions just to know if I did something wrong or I need to change some parameters.

[method_name_train.txt](https://github.com/tech-srl/code2vec/files/4502597/method_name_train.txt)
[binary_classification_train.txt](https://github.com/tech-srl/code2vec/files/4502598/binary_classification_train.txt)

Best regards,
Simone

**Comment by urialon** [comment]
Hi Simone, 
I'm sorry for the delayed response. 
You might want to increase the number of training iterations here: https://github.com/tech-srl/code2vec/blob/master/config.py#L47
Because it seems that your models keep improving over time.

However, your models are suspiciously very accurate, reaching almost 100%. You might want to verify that the dataset is not "too easy". Alternatively, maybe there are duplicates?

Best,
Uri

**Comment by urialon** [comment]
Closing due to inactivity, feel free to reopen if you have any questions.

**Comment by SimoneBrigante** [comment]
> You might want to increase the number of training iterations here: https://github.com/tech-srl/code2vec/blob/master/config.py#L47
> Because it seems that your models keep improving over time.

I have tried to increase the number of iteration, but the binary model doesn't improve, indeed it gets worse, probably because of the relatively small dataset

> However, your models are suspiciously very accurate, reaching almost 100%. You might want to verify that the dataset is not "too easy". Alternatively, maybe there are duplicates?

Yes, I was expecting to get really good results for methods name prediction, since the dataset contains simple computer science exercises, many of whom with the same name.

I have spoken with my thesis supervisor and we are really interested in the interpretability part of the model, is there a way to extract the attention arrays of many snippets or maybe understand if, in general, a particular context_path is strongly related to a certain label? 
I hope I explained myself and thanks again for your availability!

Best regards,
Simone

**Comment by urialon** [comment]
The current way to explore the attention scores is using the "predict" mode: https://github.com/tech-srl/code2vec#step-4-manual-examination-of-a-trained-model

Enabling this for a batch of examples will require some coding. If you wish to implement this, you can work with the `--test` flag that builds the computation graph here:
https://github.com/tech-srl/code2vec/blob/master/tensorflow_model.py#L124
Some of the return values that are ignored, are the attention scores.
In the `--predict` scenario, the attention weights are not ignored:
https://github.com/tech-srl/code2vec/blob/master/tensorflow_model.py#L319

So you can take these attention weights in `--test` mode, and print them or save them to a file.

Regarding further investigation of the trained model - I'm afraid that I don't have a good answer. you can try looking for interpretability papers and see if there's anything relevant...

Best,
Uri

**Comment by Anurag-Swarnim-Yadav** [comment]
@urialon 
Hi Uri,

I not able to figure out how to point code2vec or code2sec to the target file. Can you please tell me where to make changes to point it to the target file.

**Comment by urialon** [comment]
Closing due to inactivity, feel free to reopen if you have any questions.
